### PR TITLE
Optimize FP32 LM head for bf16/fp16

### PR DESCRIPTION
**Pull request by ilyasher-harmonic**
Optimize `use_fp32_lm_head` when hidden states & LM head weights are bf16 or fp16.

The previous behavior was to cast the hidden states and LM weights into new fp32 tensors and then perform a fp32xfp32 -> fp32 matmul. This PR optimizes this by skipping the fp32 cast and using a bf16xbf16 -> fp32 matmul with an fp32 accumulator to preserve the accuracy as requested by `use_fp32_lm_head`.  This saves time by removing the expensive fp32 casts, and by using a faster gemm operation.

Docs: https://docs.pytorch.org/docs/2.13/generated/torch.mm.html

### Benchmark
On H200:
Before: 5.7ms
After: 0.56ms
**10x speedup**

Benchmark code:
```python
import gc
import json

import torch


M, K, N = 40, 4096, 248320
WARMUP, REPS = 3, 10

torch.manual_seed(0)
hidden = torch.randn(M, K, device="cuda", dtype=torch.bfloat16)
weight = torch.randn(N, K, device="cuda", dtype=torch.bfloat16)


def explicit_cast():
    return torch.matmul(hidden.float(), weight.float().T)


def mm_out_dtype():
    return torch.mm(hidden, weight.T, out_dtype=torch.float32)


variants = {
    "explicit_cast_matmul": explicit_cast,
    "mm_out_dtype": mm_out_dtype,
}

results = {}
reference = None
for name, fn in variants.items():
    try:
        for _ in range(WARMUP):
            out = fn()
        torch.cuda.synchronize()
        times = []
        for _ in range(REPS):
            start = torch.cuda.Event(enable_timing=True)
            end = torch.cuda.Event(enable_timing=True)
            with torch.cuda.nvtx.range(name):
                start.record()
                out = fn()
                end.record()
            end.synchronize()
            times.append(start.elapsed_time(end))
        if reference is None:
            reference = out
            max_abs = 0.0
        else:
            max_abs = (out - reference).abs().max().item()
        results[name] = {
            "mean_ms": sum(times) / len(times),
            "min_ms": min(times),
            "max_ms": max(times),
            "max_abs_vs_explicit": max_abs,
        }
    except Exception as exc:
        results[name] = {"error": repr(exc)}
    finally:
        gc.collect()
        torch.cuda.empty_cache()

print(json.dumps({
    "shape": [M, K, N],
    "torch": torch.__version__,
    "gpu": torch.cuda.get_device_name(),
    "results": results,
}, indent=2), flush=True)

```

---
### CI States

Latest PR Test (Base): :x: [Run #31256436608](https://github.com/sgl-project/sglang/actions/runs/31256436608)
Latest PR Test (Extra): :x: [Run #31256436408](https://github.com/sgl-project/sglang/actions/runs/31256436408)

### CODE REVIEW

**Review by b8zhong — APPROVED** [review]
Good job.

**Comment by b8zhong** [comment]
@ispobock Can we merge this? I think the remaining failures are unrelated.

**Review by Fridge003 — APPROVED** [review]
